[CDAP-13267]Add implementation of the FieldLineageDataset and subscriber of FieldLineage info from TMS which writes to it.

sagarkapare opened the pull request:
JIRA: https://issues.cask.co/browse/CDAP-13267

Following changes are implemented:
1. `FieldLineageDataset` implementation to store the collection of operations and retrieve it given the specific time range. 
2. Ability to publish the FieldLineageInfo to the TMS and subscriber which retrieves it and writes to the field lineage dataset.

sagarkapare committed:
CDAP-13267 Add implementation of the FieldLineageDataset and publish the lineage graph to TMS.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataSubscriberService.java:
```diff
@@ -228,6 +247,26 @@ public void processMessage(MetadataMessage message) {
+    public void processMessage(MetadataMessage message) {
+      FieldLineageInfo info = message.getPayload(GSON, FieldLineageInfo.class);
+      ProgramRunId programRunId = message.getProgramId().run(message.getRunId());
+      fieldLineageDataset.addFieldLineageInfo(programRunId, info);
```
what happens if a message cannot be processed? Will this fail the transaction? In that case, will the TMS consumer retry processing the same message again? Can it get stuck in a fail loop?

sagarkapare replied:
Based on the current logic, if message cannot be processed, transaction fails. For messages belonging to the metadata topic, we do have retry strategy of retrying for 1000 times by default(`system.metadata.retry.policy.max.retries`). However after retrying for 1000 times, I don't see we update the messageid for ignoring the messages that are causing failures.
@chtyim is my understanding correct?

anew replied:
so it will never get past the failure?

chtyim replied:
Currently it won't skip if there is exception raised from the `processMessage`. It is expected the child class (like this one) to catch exceptions that it consider skippable and skip the message (e.g. if the MetadataMessage is considered malformed and can never be processed).

BTW, @sagarkapare I think we should also update the logic in the `AbstractMessagingSubscriberService.runOneIteration()` method, as currently it will do a no-sleeping retry after the `RetryStrategy` exhausted (meaning returning 0). We should probably default to some non-zero sleep time when that happen. I don't want to terminate the subscriber service since I don't think one subscriber bug should cause all subscribers to halt.

sagarkapare replied:
@chtyim Filed this JIRA - https://issues.cask.co/browse/CDAP-13594.. will need some discussion to better understand the problem.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/preview/PreviewDataModules.java:
```diff
@@ -92,9 +94,12 @@ protected void configure() {
         expose(LineageStoreReader.class);
 
-        bind(LineageWriter.class).to(BasicLineageWriter.class);
+        bind(LineageWriter.class).to(NoOpLineageWriter.class);
```
Did we decide that preview does not support field level lineage? Or is this future work? Is there a Jira for this? I think it would be a useful feature.

sagarkapare replied:
https://issues.cask.co/browse/CDAP-13588

anew replied:
thanks

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageReader.java:
```diff
@@ -0,0 +1,84 @@
+   *
+   * @param endPoint the EndPoint for which the fields need to be returned
+   * @param start start time in milliseconds
+   * @param end end time in milliseconds
```
is the end time included or excluded? Please document that

sagarkapare replied:
done.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageReader.java:
```diff
@@ -0,0 +1,84 @@
+   * the specified EndPointField.
+   *
+   * @param endPointField the EndPointField for which incoming summary to be returned
+   * @param start start time in milliseconds
```
ditto

sagarkapare replied:
done

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageReader.java:
```diff
@@ -0,0 +1,84 @@
+   *
+   * @param endPointField the EndPointField for which outgoing summary to be returned
+   * @param start start time in milliseconds
+   * @param end end time in milliseconds
```
ditto

sagarkapare replied:
done.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageReader.java:
```diff
@@ -0,0 +1,84 @@
+   *
+   * @param endPoint the EndPoint for which incoming operations are to be returned
+   * @param start start time in milliseconds
+   * @param end end time in milliseconds
```
ditto

sagarkapare replied:
done.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageReader.java:
```diff
@@ -0,0 +1,84 @@
+   *
+   * @param endPoint the EndPoint for which outgoing operations are to be returned
+   * @param start start time in milliseconds
+   * @param end end time in milliseconds
```
ditto

sagarkapare replied:
done.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageReader.java:
```diff
@@ -0,0 +1,84 @@
+ */
+public interface FieldLineageReader {
+  /**
+   * Get set of fields generated for the specified EndPoint in a given time range.
```
what does "generated" mean? Are these the fields that are written/produced/computed to the endpoint by write operations recorded in the field lineage over this time range?

sagarkapare replied:
updated the documentation. please take a look again.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageReader.java:
```diff
@@ -0,0 +1,84 @@
+
+  /**
+   * Get the incoming summary for the specified EndPointField in a given time range.
+   * Incoming summary consists of set of EndPointFields which were responsible for generating
```
similar here. generating is not clear

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageReader.java:
```diff
@@ -0,0 +1,84 @@
+
+  /**
+   * Get the outgoing summary for the specified EndPointField in a given time range.
+   * Outgoing summary consists of set of EndPointFields which were generated by the
```
I guess I am irritated by the term "generated by the fields". Perhaps you mean generated (or better "produced", "computed"?) from these fields by operations in the field lineage?

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageReader.java:
```diff
@@ -0,0 +1,84 @@
+  Set<EndPointField> getOutgoingSummary(EndPointField endPointField, long start, long end);
+
+  /**
+   * Get the set of operations which were responsible for generating the fields
```
perhaps better, which participated in the computation of the fields of the endpoint. However, I thought this is for a single field?

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageReader.java:
```diff
@@ -0,0 +1,84 @@
+
+  /**
+   * Get the set of operations which were performed on the fields of the specified EndPoint
+   * to generate fields of the downstream EndPoints. Along with the operations, program
```
You use the term "generate" everywhere. perhaps that is ok if you add an explanation of what it means in the Javadoc for the class.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageReader.java:
```diff
@@ -0,0 +1,84 @@
+  Set<ProgramRunOperations> getIncomingOperations(EndPoint endPoint, long start, long end);
+
+  /**
+   * Get the set of operations which were performed on the fields of the specified EndPoint
```
Again I thought we compute this for a single field of the endpoint? The UI will always show the operations for a single field, right?

sagarkapare replied:
yes. thats correct. after we get set of operations, we will need to do post processing, to create operations for single field. That will come in as a part of next PR which will contain REST api.

anew replied:
why not perform the projection here? It seems straight-forward

anew replied:
as mentioned above I don't see why we first compute more than needed and then project it back to a smaller set. Why not limit it right away to only the operations of interest?

sagarkapare replied:
Have added logic to get projected operations in the last commit. Unfortunately this increased the PR size though by few hundred lines.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/ProgramRunOperations.java:
```diff
@@ -0,0 +1,68 @@
+
+/**
+ * Represents the set of program runs along with the set of field operations
+ * performed by each of the program run in that set.
```
Represents a set of program runs that performed the same set of field operations.

sagarkapare replied:
done.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java:
```diff
@@ -69,6 +69,12 @@
   // combination of (EndPoint,Field) which were responsible for generating it.
   private transient Map<EndPointField, Set<EndPointField>> incomingSummary;
 
+  // Source endpoints in the lineage info
```
I am a little confused by what is transient and what is not. It seems that ```writeOperations``` and ```operationMap``` are redundant because they are trivial to compute from ```operations```. But they all get serialized.

Whereas destinationFields, outgoingSummary, and incomingSummary are not trivial to compute. Why are they transient?

sagarkapare replied:
made `destinationFields`, `outgoingSummary`, and `incomingSummary` non transient.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/FieldLineageWriter.java:
```diff
@@ -0,0 +1,34 @@
+import co.cask.cdap.proto.id.ProgramRunId;
+
+/**
+ * Defines an interface to write the field lineage information for a program run.
```
Should this mention that write() will be called without a transaction and it must start its own tx?

sagarkapare replied:
done.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/MessagingLineageWriter.java:
```diff
@@ -66,13 +72,25 @@ public void addAccess(ProgramRunId programRunId, StreamId streamId,
     try {
       Retries.callWithRetries(() -> messagingService.publish(request), retryStrategy, Retries.ALWAYS_TRUE);
     } catch (Exception e) {
-      throw new RuntimeException("Failed to publish data access lineage: " + lineage, e);
```
maybe better to pass in the type of lineage that is being written, to keep the error messages on point.

sagarkapare replied:
type is part of message already - https://github.com/caskdata/cdap/blob/develop/cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/MetadataMessage.java#L123

anew replied:
So the error message will contain the entire message? That might be overwhelming. Can we just have a summary in the exception, and LOG.trace the entire message?

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java:
```diff
@@ -0,0 +1,357 @@
+  //
+  // c) Name: o | <EndPoint> | <field>
+  // For each field belonging to each of the source EndPoint, this column stores the
+  // collection of destination EndPoint and fields that were generated by the given field.
```
generated from.

sagarkapare replied:
done.

anew replied:
or better, computed from?

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java:
```diff
@@ -0,0 +1,357 @@
+    addFieldLineageInfoReferenceRecords(programRunId, info);
+  }
+
+  /**
```
Can you move the addSummary() method to here? It's easier to find near where it is being called

sagarkapare replied:
done.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java:
```diff
@@ -0,0 +1,357 @@
+      addOperationReferenceRecord(OUTGOING_DIRECTION_MARKER, source, programRunId, info.getChecksum());
+    }
+  }
+
```
and move addOperationReferenceRecord() here? thanks

sagarkapare replied:
done.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java:
```diff
@@ -0,0 +1,357 @@
+    for (long checksum : checksums) {
+      byte[] rowKey = getChecksumRowKey(checksum);
+      String value = Bytes.toString(table.get(rowKey, columnKey));
+      Set<String> fields = GSON.fromJson(value, SET_FIELD_TYPE);
```
Can GSON read directly from the bytes? That would be more efficient that converting to string first. It matters here because this may be called for many checksums, depending on the time range.

anew replied:
You did not address this comment. This can be a very large string and it would be good to avoid constructing it

sagarkapare replied:
Filed this JIRA for the optimization - https://issues.cask.co/browse/CDAP-13595

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java:
```diff
@@ -0,0 +1,357 @@
+    for (long checksum : checksums) {
+      byte[] rowKey = getChecksumRowKey(checksum);
+      String value = Bytes.toString(table.get(rowKey, columnKey));
+      Set<String> fields = GSON.fromJson(value, SET_FIELD_TYPE);
```
Can it ever happen that this throws a JSON syntax exception? Should we catch that, log a warning and ignore the record?

sagarkapare replied:
shouldn't happen since the JSON is created by the platform itself.. but added checks anyway

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java:
```diff
@@ -0,0 +1,357 @@
+
+    for (long checksum : checksums) {
+      byte[] rowKey = getChecksumRowKey(checksum);
+      String value = Bytes.toString(table.get(rowKey, columnKey));
```
can this ever be null?

anew replied:
ok looks like you cover that below.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java:
```diff
@@ -0,0 +1,357 @@
+      String value = Bytes.toString(table.get(rowKey, columnKey));
+      Set<EndPointField> endPointFields = GSON.fromJson(value, SET_ENDPOINT_FIELD_TYPE);
+      if (endPointFields != null) {
+        // TODO: can it be null ever??
```
hehe yes if the above is null

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java:
```diff
@@ -0,0 +1,357 @@
+    Set<Long> checksums = checksumsWithProgramRunsInRange.keySet();
+    Set<ProgramRunOperations> result = new HashSet<>();
+
+    for (long checksum : checksums) {
```
better iterate over the entries than iterating over the keys and doing a get for each key

sagarkapare replied:
fixed.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java:
```diff
@@ -0,0 +1,357 @@
+    for (long checksum : checksums) {
+      byte[] rowKey = getChecksumRowKey(checksum);
+      String value = Bytes.toString(table.get(rowKey, RAW_OPERATION_MARKER));
+      Set<Operation> operations = GSON.fromJson(value, SET_OPERATION_TYPE);
```
this adds all operations in the graph. But we only need the operations that influence the given endpoint. Actually (see my other comment) only one field of that endpoint.

sagarkapare replied:
Added functionality in the last commit - https://github.com/caskdata/cdap/pull/10213/commits/5caf434c48115fd8933cbafe5d7181e0de4ffff9

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java:
```diff
@@ -0,0 +1,357 @@
+      Row row;
+      while ((row = scanner.next()) != null) {
+        if (LOG.isTraceEnabled()) {
+          LOG.trace("Got row key = {}", Bytes.toString(row.getRow()));
```
If you already log this, can you also log what you got this for? direction, endpoint, time range.

sagarkapare replied:
fixed.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java:
```diff
@@ -0,0 +1,357 @@
+
+  private byte[] getScanStartKey(byte[] direction, EndPoint endPoint, long end) {
+    // time is inverted, hence we need to have end time in start key.
+    // Since end time is exclusive, add 1 to make it inclusive.
```
this comment confuses me. Why make it inclusive?

sagarkapare replied:
removed this method and added explanation in the following comment.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java:
```diff
@@ -0,0 +1,357 @@
+
+  private byte[] getScanEndKey(byte[] direction, EndPoint endPoint, long start) {
+    // time is inverted, hence we need to have start time in end key.
+    // Since start time is inclusive, subtract 1 to make it exclusive.
```
this, too, confuses me.

sagarkapare replied:
removed these methods.. we only need to add 1 always to the inverted time. the reason is
once the time is inverted, original end-time becomes the start time for the hbase scan, however we should not return the record corresponding to this start time (inverted end-time) since end time is exclusive, so we add 1. Similarly once the time is inverted, original start-time becomes the end time of the scan. Since we want to return the records corresponding to this end time (inverted start-time), we add 1.

anew commented on cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDatasetTest.java:
```diff
@@ -0,0 +1,180 @@
+                                           DatasetProperties.EMPTY, null);
+  }
+
+  private List<Operation> getOperations(String[] readOutput) {
```
this is more a generateOperations() than a getter

sagarkapare replied:
renamed.

anew commented on cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDatasetTest.java:
```diff
@@ -0,0 +1,180 @@
+  @ClassRule
+  public static DatasetFrameworkTestUtil dsFrameworkUtil = new DatasetFrameworkTestUtil();
+
+  @Test
```
I think this could use more test cases. You need to test 
- that the time ranges work as designed (start inclusive, end exclusive)
- edge cases such as cyclic graphs, graphs where the writes or the reads are not connected, etc
- merging of the field summaries from multiple runs
- etc.

sagarkapare replied:
Added tests for time range and merge of field summaries. Have already filed JIRA - https://issues.cask.co/browse/CDAP-13548 for handling cases of cycles, will add tests for it as a part of that JIRA.

anew reviewed (commented):
Please see my comments

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageReader.java:
```diff
@@ -25,59 +26,60 @@
-   * @param start start time in milliseconds
-   * @param end end time in milliseconds
-   * @return set of fields generated for a given EndPoint
+   * @param start start time(inclusive) in milliseconds
```
blank before (

sagarkapare replied:
fixed.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageReader.java:
```diff
@@ -25,59 +26,60 @@
-   * @param end end time in milliseconds
-   * @return set of fields generated for a given EndPoint
+   * @param start start time(inclusive) in milliseconds
+   * @param end end time(exclusive) in milliseconds
```
here, too

sagarkapare replied:
fixed.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java:
```diff
@@ -149,6 +151,9 @@ public FieldLineageInfo(Collection<? extends Operation> operations) {
 
     this.operations = new HashSet<>(operations);
     this.checksum = computeChecksum();
+    this.destinationFields = computeDestinationFields();
```
computing these can be expensive. Perhaps they should be computed on demand only if required? That means, however, need to make sure they are computed before serializing for storing it.

anew reviewed (commented):
I had a few more comments.

Also, please clarify:
- Error handling in subscriber - do we need fix anything here?
- Can we Inline the projection of operations to the subgraph for one destination field into the method that collects the operations?
- How large do exception messages become, when they contain the entire TMS message? Is there a need for improvement?

Thanks

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java:
```diff
@@ -0,0 +1,379 @@
+    table.put(put);
+  }
+
+  public Set<String> getFields(EndPoint endPoint, long start, long end) {
```
All public methods should have javadoc to clearly state the method contract. For non-public method, it would be good to have javadoc as well for non-trivial ones.

sagarkapare commented:
@anew Have addressed the comments. Please take a look again. 
Specifically - 
1. Error handling in subscriber - do we need fix anything here? 
>> Filed this JIRA https://issues.cask.co/browse/CDAP-13594
2. Can we Inline the projection of operations to the subgraph for one destination field into the method that collects the operations?
>> Added implementation in last commit.
3. How large do exception messages become, when they contain the entire TMS message? Is there a need for improvement?
>> Logging the entire message at TRACE level and only surfacing the type and program run id information in the exception as suggested.

thanks!

anew commented on cdap-api/src/main/java/co/cask/cdap/api/lineage/field/Operation.java:
```diff
@@ -16,6 +16,8 @@
 
 package co.cask.cdap.api.lineage.field;
 
+import java.util.Objects;
```
why is there a new import but no new code that uses it?

sagarkapare replied:
using it now.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/DefaultFieldLineageReader.java:
```diff
@@ -89,22 +93,42 @@ public DefaultFieldLineageReader(DatasetFramework datasetFramework, TransactionS
+      try {
+        FieldLineageInfo info = new FieldLineageInfo(programRunOperation.getOperations());
+        Set<Operation> fieldOperations = incoming ?
+                info.getIncomingOperationsForField(new EndPointField(endPoint, field))
```
this is still computing all operation for the endpoint, then removing the ones that are not for the field in question. Can't we filter already in  fieldLineageDataset.getIncomingOperations/getOutgoingOperations?

sagarkapare replied:
We store all operations as column in dataset. I wanted to minimize the work we do inside transaction (to avoid timeouts). So dataset simply returns all operations without doing any processing. FieldLineageReader then does the post processing on them to return the subset. Please let me know if this make sense.

anew replied:
yes that makes sense.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java:
```diff
@@ -300,6 +350,215 @@ private void computeIncomingSummaryHelper(EndPointField destination, Operation c
+
+    if (currentOperation.getType() == OperationType.TRANSFORM) {
+      TransformOperation transform = (TransformOperation) currentOperation;
+      operationToAdd = new TransformOperation(transform.getName(), transform.getDescription(), inputFields,
```
why remove some inputs?

sagarkapare replied:
removed this code.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java:
```diff
@@ -300,6 +350,215 @@ private void computeIncomingSummaryHelper(EndPointField destination, Operation c
+      Operation operation = result.get(operationToAdd.getName());
+      if (operation != null) {
+        // operation was already added as a part of another path
+        // create new operation after merging the input and output fields of both
```
I am not sure I understand this... why do we have to merge it with itself? If it has already been added, then what more is there to do?

sagarkapare replied:
removed this code per discussion.

anew reviewed (commented):
I think I need to understand what exactly the outgoing or incoming operation are for a field. I thought it is the subset of all operations that contributes to the field, or were influenced by the field. Isn't that computed by a simple depth-first traversal of the operations graph?

sagarkapare commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/DefaultFieldLineageReader.java:
```diff
@@ -118,14 +117,11 @@ public DefaultFieldLineageReader(DatasetFramework datasetFramework, TransactionS
         Set<Operation> fieldOperations = incoming ?
                 info.getIncomingOperationsForField(new EndPointField(endPoint, field))
-                : info.getOutgoingOperationsFromField(new EndPointField(endPoint, field));
+                : Collections.EMPTY_SET;
```
Returning EMPTY_SET for outgoing operations since the work is not planned for 5.0. Will add implementation later if required.

sagarkapare commented:
@anew Addressed comments. please take a look again. thanks!

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java:
```diff
@@ -487,6 +487,36 @@ private Operation mergeOperationInputOutputs(Operation first, Operation second)
+          continue;
+        }
+
+        Map<EndPointField, Set<EndPointField>> result = new HashMap<>();
```
is ```result``` still needed? Looks like ```computeIncomingSummaryHelper``` only adds to it, and here it is discarded without reading it.

anew commented on cdap-api/src/main/java/co/cask/cdap/api/lineage/field/ReadOperation.java:
```diff
@@ -68,4 +69,25 @@ public EndPoint getSource() {
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
```
these first two if's are already taken care of by super.equals()

anew replied:
revised. The first one should be checked in each subclass, but is not needed in Operation because that is abstract.

sagarkapare replied:
fixed.

anew commented on cdap-api/src/main/java/co/cask/cdap/api/lineage/field/TransformOperation.java:
```diff
@@ -67,4 +68,25 @@ public TransformOperation(String name, String description, List<InputField> inpu
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
```
same here

anew commented on cdap-api/src/main/java/co/cask/cdap/api/lineage/field/WriteOperation.java:
```diff
@@ -67,4 +68,25 @@ public EndPoint getDestination() {
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
```
and here

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java:
```diff
@@ -0,0 +1,429 @@
+                Bytes.toString(RAW_OPERATION_MARKER)));
+      }
+      if (operations != null) {
+        result.add(new ProgramRunOperations(entry.getKey(), checksumsWithProgramRunsInRange.get(entry.getKey()),
```
isn't ```checksumsWithProgramRunsInRange.get(entry.getKey())``` the same as ```entry.getValue()```?

sagarkapare replied:
fixed.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java:
```diff
@@ -0,0 +1,429 @@
+        } catch (JsonSyntaxException e) {
+          LOG.warn(String.format("Failed to parse json from row '%s' and column '%s'.", Bytes.toString(row.getRow()),
+                  Bytes.toString(PROGRAM_MARKER)));
+        }
```
better ```continue;``` here, then you don't need the ```if (programRunId != null)``` below.

sagarkapare replied:
done

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java:
```diff
@@ -282,7 +283,7 @@ private long computeChecksum() {
       List<InputField> inputs = write.getInputs();
       for (InputField input : inputs) {
-        Set<String> visitedOperationNames = new HashSet<>();
+        Set<String> visitedOperationNames = new HashSet<>(Collections.singletonList(write.getName()));
```
why create a singleton list for this? Better to call ```visitedOperationNames.add(write.getName())```;

sagarkapare replied:
done

anew reviewed (commented):
I have a few more comments, overall looks great now.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java:
```diff
@@ -0,0 +1,429 @@
+        try {
+          programRunId = GSON.fromJson(Bytes.toString(row.get(PROGRAM_MARKER)), ProgramRunId.class);
+        } catch (JsonSyntaxException e) {
+          LOG.warn(String.format("Failed to parse json from row '%s' and column '%s'.", Bytes.toString(row.getRow()),
```
I think the message should mention that this run is being ignored.

sagarkapare replied:
updated the message

sagarkapare commented:
@anew Please take a look again. thanks!

sagarkapare commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/ProgramRunOperations.java:
```diff
@@ -26,20 +26,14 @@
 
-  public ProgramRunOperations(long checksum, Set<ProgramRunId> programRunIds, Set<Operation> operations) {
-    this.checksum = checksum;
+  public ProgramRunOperations(Set<ProgramRunId> programRunIds, Set<Operation> operations) {
```
Missed this change in my last commit. Removed checksum from ProgramRunOperations, since its not needed.

sagarkapare commented:
Build is running here - https://builds.cask.co/browse/CDAP-DUT6480-7

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java:
```diff
@@ -375,10 +376,10 @@ private void computeIncomingSummaryHelper(EndPointField destination, Operation c
         computeIncomingSummaryHelper(new EndPointField(write.getDestination(), input.getName()),
-                operationsMap.get(input.getOrigin()), write, visitedOperationNames, result);
+                                     operationsMap.get(input.getOrigin()), write, visitedOperationNames,
+                                     new HashMap<>());
```
what I meant by my previous comment is that this parameter is not needed. The method computeIncomingSummaryHelper() only adds entries to it, but nobody ever reads from it. Or am I missing something?

sagarkapare replied:
`computeIncomingSummaryHelper()` also gets called while computing incoming summary from line 288 above, which uses the HashMap. This call just uses the same helper function to reuse the recursion logic.

anew replied:
Ah ok I did not see that in the diff. All good.

anew reviewed (commented):
I only had one comment left. Thanks -

anew commented:
LGTM

sagarkapare commented:
Build https://builds.cask.co/browse/CDAP-DUT6498-3 green for PR #10231 which was on the top of this. Merging.
